src/game.cpp

```diff
@@ -6601,109 +6616,298 @@ void game::print_visibility_info( const catacurses::window &w_look, int column,
+                col = c_yellow;
+            } else if( cur_hp >= max_hp * 0.1 ) {
+                tempStr = _( "It is severely injured." );
+                col = c_light_red;
+            } else {
+                tempStr = _( "It is nearly dead!" );
+                col = c_red;
+            }
+            info.push_back( trim_to_width( colorize( tempStr, col ), max_width ) );
+            //description
+            foldstring( info, colorize( mon->type->get_description(), c_white ), max_width );
+
+            return info.size() - old_size;
+        }
+        //NPC INFO
+        auto *npc_ = dynamic_cast<const npc *>( creature );
+        if( npc_ ) {
+            //name
+            tempStr = string_format( _( "NPC: %s" ), npc_->name );
+            info.push_back( trim_to_width( colorize( tempStr, c_white ), max_width ) );
+            //wielding
+            if( npc_->is_armed() ) {
+                tempStr = string_format( _( "Wielding a %s" ), npc_->weapon.tname() );
+                info.push_back( trim_to_width( colorize( tempStr, c_red ), max_width ) );
+            }
+            //wearing
+            std::string wornStr;
+            for( auto &it : npc_->worn ) {
+                wornStr += it.tname() + ", ";
+            }
+            if( wornStr.size() > 1 ) {
+                wornStr.resize( wornStr.size() - 2 ); // remove the last ", "
+            }
+            if( !wornStr.empty() ) {
+                tempStr = _( "Wearing: " ) + remove_color_tags( wornStr );
+                foldstring( info, colorize( tempStr, c_blue ), max_width );
+            }
+            //mutations
+            const int visibility_cap = g->u.get_per() - rl_dist( g->u.pos(), npc_->pos() );
+            const auto trait_str = npc_->visible_mutations( visibility_cap );
+            if( !trait_str.empty() ) {
+                tempStr = _( "Traits: " ) + remove_color_tags( trait_str );
+                foldstring( info, colorize( tempStr, c_green ), max_width );
+            }
+
+            return info.size() - old_size;
+        }
+        //PLAYER INFO
+        auto *you = dynamic_cast<const player *>( creature );
+        if( you == &u ) {
+            info.push_back( trim_to_width( string_format( _( "You (%s)" ), you->name ), max_width ) );
+        }
     }
+
+    return info.size() - old_size;
 }
 
-void game::print_vehicle_info( const vehicle *veh, int veh_part, const catacurses::window &w_look,
-                               const int column, int &line, const int last_line )
+size_t game::vehicle_info( std::vector<std::string> &info, const vehicle *veh, int veh_part,
+                           const int max_width )
 {
+    size_t old_size = info.size();
+
     if( veh ) {
-        mvwprintw( w_look, ++line, column, _( "There is a %s there. Parts:" ), veh->name.c_str() );
-        line = veh->print_part_list( w_look, ++line, last_line, getmaxx( w_look ), veh_part );
+        //brief description
```

Comment:
Same as above, but using the relevant vehicle functions.

Reply:
I agree, I didn't touched those 4 (monster, npc, player, vehicle) because I didn't wanted to span on too many cpp, but that's the solution I would have preferred as well. 
I'll do tomorrow 👍

Reply:
If you want the `get_info` functions to return different values, then you should pass a enum bitfield into `get_info` that defines what parts to omit or include.  But the `get_info` functions should be class functions.

So `std::vector<std:string> veh_display = veh->get_info( max_x, vehicles::omit_fuel | vehicles::omit_cargo );` would fill `veh_display` with the folded vehicle strings for whatever a vehicle displays except the fuel and cargo data.

It's pretty reasonable for the vehicle code to know what it's fuel or cargo data is, so this is the appropriate level of abstraction.

Reply:
> if you add up all the possible variation people might want to have

This MUST be limited, no matter what you do it's a nightmare to support.

Generating the various fields in individual methods is a good idea, but those methods should belong to the classes themselves, not helpers.

If you want to break down stringification of classes to various individual fields, feel free to start breaking them down into a number of helpers on the classes that return just e.g. the stringified fuel gague, the stringified weight, etc...   Helpers that live in their own class that know how to stringify various pieces of other classes are an antipattern that exposes too much of each classes internal data.

Reply:
After more thonking I realized I agree, as long as the scenario of the monolithic switch function doesn't get to occur, which I'm afraid it is what would happen with the enum flag solution.
I guess what I was thinking was relying heavily on the fact the vehicle data type (and the others) are well known types and they will not change, but that was a big assumption :\

Reply:
Generally, the get_info() function is going to be of the form:
```
if field_1_option_a
   ret_vector.pushback( field_1_format_a )
else if field_1_option_b
   ret_vector.pushback( field_1_format_b )

if field_2_option_a
   ret_vector.pushback( field_2_format_a )
else if field_2_option_b
   ret_vector.pushback( field_2_format_b )
```
and so forth.

But seriously, stop overthinking the design and worrying about what happens two iterations down the road.  Produce a good enough solution for your current problems that doesn't obviously do dumb things or cause problems in the future.

The vehicle class, for instance, has an entire file full of class functions for displaying stuff (`vehicle_display.cpp`) so that it's out of my hair 95% of the time.  But those are still class functions so they have access to class members and aren't leaking the abstraction.

Reply:
I can see that being nice, if each field is it's own function (and by field I mean the smallest piece of information, like the fuel, or the cargo). So even if fuel has 5 different formats, a switch statement still fit in my screen, so I like that.
Now the question is about enum, I would prefer to use "enum class" to avoid people passing plain integers into this functions, which kills readability.
Is that ok? Or there are even stronger reason for wanting a plain enum?

Reply:
well not that it matters just yet, since there is only one format for things, so might as well do without